### PR TITLE
Implementing "collect evidence" mechanic

### PR DESCRIPTION
**Pull request by theelk801**
This should be good to go now, I'm not sure what tests to write for it right now though since it's mostly about selecting targets.

### CODE REVIEW

**Review by xenohedron — COMMENTED** [review]
I think a basic unit test is appropriate here.

Code looks good enough for now, need to make a new issue for min target handling in both this and CrewCost (also applies to Mossbridge Troll and Phyrexian Dreadnought) as a future refactor

File: Mage/src/main/java/mage/abilities/costs/common/CollectEvidenceCost.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        player.choose(Outcome.Exile, target, source, game);
+        Cards cards = new CardsImpl(target.getTargets());
+        paid = cards
+                .getCards(game)
+                .stream()
+                .filter(Objects::nonNull)
+                .mapToInt(MageObject::getManaValue)
+                .sum() >= amount;
+        if (paid) {
+            player.moveCards(cards, Zone.EXILED, source, game);
```
[the remark this answers is not in the document]
Comment (theelk801) [reply Mage/src/main/java/mage/abilities/costs/common/CollectEvidenceCost.java:84]: I don't think there's any need to check

**Comment by theelk801** [comment]
I would write a test but I'm not sure what the test would actually be since it involves targeting and making choices which is hard to actually test in the framework

**Comment by xenohedron** [comment]
Unit test suggestion:

check that cost is paid with possible combinations:
- one exiled card equal to number
- one exiled card greater than number
- two exiled cards, total equal to number
- three exiled cards, with the first two total greater than number

check that the right cards are exiled after cost is paid

check that Fuming Effigy has one trigger even with multiple cards

future: check "if evidence was collected" for both paid and not paid

File: Mage/src/main/java/mage/abilities/keyword/CollectEvidenceAbility.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        if (!(ability instanceof SpellAbility)) {
+            return;
+        }
+
+        Player player = game.getPlayer(ability.getControllerId());
+        if (player == null) {
+            return;
+        }
+
+        this.resetBargain();
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/keyword/CollectEvidenceAbility.java:78]: this should be renamed

File: Mage/src/main/java/mage/abilities/keyword/CollectEvidenceAbility.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        boolean canPay = additionalCost.canPay(ability, this, ability.getControllerId(), game);
+        if (!canPay || !player.chooseUse(Outcome.Sacrifice, promptString, ability, game)) {
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/keyword/CollectEvidenceAbility.java:80]: should be different outcome

**Review by xenohedron — COMMENTED** [review]
still pending unit tests, implementations look good so far though

File: Mage/src/main/java/mage/abilities/keyword/CollectEvidenceAbility.java (hunk 2 above)
Comment (theelk801) [reply Mage/src/main/java/mage/abilities/keyword/CollectEvidenceAbility.java:78]: missing a rename is possibly the most embarrassing thing that can happen

**Comment by theelk801** [comment]
I'm doing the tests now

**Review by xenohedron — APPROVED** [review]
minor UI discrepancy, not sure if you can get the number into the prompt for `CollectEvidenceAbility` ? It appears fine for the `DoIfCostPaid`:
image

image

that's all I could find, good work

**Pull request merged**